java/vector/src/main/java/org/apache/arrow/vector/dictionary/Dictionary.java

```diff
@@ -72,4 +74,14 @@ public boolean equals(Object o) {
   public int hashCode() {
     return Objects.hash(encoding, dictionary);
   }
+
+  @Override
+  public int mark() {
+    return reset == true ? 0 : dictionary.getValueCount();
+  }
+
+  @Override
+  public void reset() {
+    reset = true;
+  }
```

Comment:
I guess what makes me uneasy overall is that we have introduced mutable state in the Dictionary itself to track things needed for writing, and I think instead it would be better to try and keep that state contained in the writer itself as it was before. Also, I agree that 'mark' and 'reset' are somewhat confusing names, though I suppose they make sense in the context of the writer.

Reply:
Yeah, that's fair, the state is ugly. I'm trying to keep the API changes to a minimum for backwards compatibility and make it easy on devs to write multiple batches in the current idiom. 

Taking CPP as the canonical implementation, their writer has a method to write replacement or delta dictionaries before a batch but the Java API assumes immutable dictionaries are provided at writer creation via the `DictionaryProvider`. 

We _could_ add an `DictionaryProvider.update(Dictionary dictionary)` method that mimics the CPP behavior (and have a dictionary builder helper like `builder_dict`) but I'd still need a bit of state in the provider to determine when a batch write was performed. Without that state, the only way I know of to determine if a dictionary has changed and should be flushed in a batch is to hash on the _contents_ of the dictionary vector on every batch write. That would really hurt performance with large dictionaries. 

What do you think?

Reply:
I'm more comfortable with adding it to the provider. You could even introduce it as a subclass or wrapper/proxy class that the writer checks for? Then the base provider doesn't support replacement/deltas and the new provider would let you explicitly update dictionaries.

I also don't like Arrow-Java's tendency towards mutable state/data. It makes things like this messy since we can't rely on object identity, unfortunately.

Reply:
Ack, I'll re-work it as a subclass, thanks! Yup, immutable objects would be nice for this case.